### PR TITLE
New admin states restrictions.

### PR DESCRIPTION
**Pull request by jaleszek**
1. Admin can't change any state before the submission end
2. Admin can't change the state if `not_submitted`
3. Admin has restrictions to relevant states after the deadline and if app is submitted ?

### CODE REVIEW

File: app/models/form_answer_state_machine.rb [hunk 1]
```diff
@@ -56,14 +56,10 @@ def automatic_states
...
-    if subject.is_a?(Admin)
-      STATES
-    else
-      permitted_states_with_deadline_constraint - automatic_states
-    end
+    permitted_states_with_deadline_constraint# - automatic_states
   end
 
-  def perform_transition(state, subject = nil)
+  def perform_transition(state, subject = nil, validate = true)
```
Comment (matthewford) [inline app/models/form_answer_state_machine.rb]: Assignment Branch Condition size for perform_transition is too high. [19.72/15]

File: app/models/form_answer_state_machine.rb [hunk 2]
```diff
@@ -56,14 +56,10 @@ def automatic_states
   end
 
   def collection(subject)
-    if subject.is_a?(Admin)
-      STATES
-    else
-      permitted_states_with_deadline_constraint - automatic_states
-    end
+    permitted_states_with_deadline_constraint# - automatic_states
```
Comment (matthewford) [inline app/models/form_answer_state_machine.rb]: Put a space before an end-of-line comment.

File: spec/models/form_answer_state_machine_spec.rb [hunk 3]
```diff
@@ -7,6 +7,17 @@
     context "deadline expired" do
       let!(:settings) { create(:settings, :expired_submission_deadlines) }
 
+      describe "application_in_progress -> withdrawn" do
```
Comment (dreamfall) [inline spec/models/form_answer_state_machine_spec.rb]: `assessment_in_progress`?

**Comment by jaleszek** [comment]
@dreamfall fixed

File: spec/features/admin/form_answers/states_transitions_spec.rb [hunk 4]
```diff
@@ -3,23 +3,25 @@
...
+  let!(:form_answer) { create(:form_answer, state: "assessment_in_progress") }
 
   before do
+    setting = Settings.current_submission_deadline
+    setting.update(trigger_at: DateTime.now - 1.hour)
     login_admin(admin)
     visit admin_form_answer_path(form_answer)
   end
 
   it "changes the state from application_to_progress to submitted" do
```
Comment (dreamfall) [inline spec/features/admin/form_answers/states_transitions_spec.rb]: I guess it's better to change title here too.

**Comment by dreamfall** [comment]
looks good

**Comment by RuslanHamidullin** [comment]
This one also checked

**Pull request merged**